### PR TITLE
Use an op counter to decide when to realize a kernel

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117030

Instead of checking the number of bytes in the string representation
of the kernel

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by isuruf** [comment]
@peterbell10, @lezcano, it looks like the bytes based heuristic is roughly equal to 30 operations.

**Comment by lezcano** [comment]
We can benchmark this see how it fares, and as discussed, I'm not opposed to landing it, as it's a more sane heuristic than looking at the string, but https://github.com/pytorch/pytorch/pull/116959#discussion_r1445788366 still stands.
It'd be nice to end up with a finer algorithm for this, perhaps trying to get a rough estimate of the number of registers used by a kernel.

**Comment by isuruf** [comment]
I ran the benchmarks on [a1ecf77](https://github.com/pytorch/pytorch/pull/117030/commits/a1ecf7787ded29f426f9dc2475d68992a1db6427). Is there a way to add a permalink here to the dashboard?

**Comment by lezcano** [comment]
It is possible, yes. Otherwise putting up a screenshot should also work

**Comment by isuruf** [comment]
Just found the button. [Here](https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2009%20Jan%202024%2016%3A21%3A38%20GMT&stopTime=Tue%2C%2016%20Jan%202024%2016%3A21%3A38%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/isuruf/22/head&lCommit=a1ecf7787ded29f426f9dc2475d68992a1db6427&rBranch=main&rCommit=d78776e2e60ae0257847097dda3eb2fdc6f32476)'s the link

**Comment by isuruf** [comment]
Yes, sorry. [Correct link](https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2009%20Jan%202024%2016%3A21%3A38%20GMT&stopTime=Tue%2C%2016%20Jan%202024%2016%3A21%3A38%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/isuruf/22/head&lCommit=a1ecf7787ded29f426f9dc2475d68992a1db6427&rBranch=main&rCommit=d78776e2e60ae0257847097dda3eb2fdc6f32476)

**Comment by isuruf** [comment]
I think those are noise. For eg see [link](https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2009%20Jan%202024%2016%3A21%3A38%20GMT&stopTime=Tue%2C%2016%20Jan%202024%2016%3A21%3A38%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/isuruf/22/head&lCommit=a1ecf7787ded29f426f9dc2475d68992a1db6427&rBranch=main&rCommit=d78776e2e60ae0257847097dda3eb2fdc6f32476) and the values are almost the same.

**Review by lezcano — APPROVED** [review]
Cool. As discussed, this is alright and a good step forward, but it'd be even better to completely remove this heuristic and have one more based on read/writes and an estimation of number of registers or so. Or even remove this heuristic altogether? Either that or at least tune this `= 30` number.

In any case, this is a good step forward.

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -393,13 +418,32 @@ def _index(ranges, prefix="i"):
...
+            self.inner_fn, *self.inner_fn_args()
+        )
+
     @cache_on_self
     def inner_fn_str_len(self):
         return len(self.inner_fn_str())
 
-    def inner_fn_str(self):
-        index = self._index(self.ranges)
-        return V.KernelFormatterHandler.ir_to_string(self.inner_fn, index)
```
Comment (lezcano) [inline torch/_inductor/ir.py:402]: Now `ir_to_string` is not used anywhere, we can probably remove it.

Comment (peterbell10) [reply torch/_inductor/ir.py:402]: `inner_fn_str` still exists and is used to print the IR node during error messages.

**Comment by isuruf** [comment]
> As discussed, this is alright and a good step forward, but it'd be even better to completely remove this heuristic and have one more based on read/writes and an estimation of number of registers or so. Or even remove this heuristic altogether? Either that or at least tune this = 30 number.

Both are good suggestions. I will open an issue about it and ping others who worked on this for insights.

**Comment by lezcano** [comment]
Does this = 30 unblock the masked PR? Or do we still hit the same issue?
At any rate, it'd be good to put up a PR and a bench to see if simply removing this heuristic is a viable option.

**Comment by isuruf** [comment]
Yes, it does unblock the avg_pool PR. I don't think simply removing the heuristic is a viable option. Increasing the threshold resulted in failures at https://github.com/pytorch/pytorch/pull/116692. (As discussed the test case might be a good candidate for xfail though). Looking at https://github.com/pytorch/pytorch/blob/c30346db0e89a58b6ab7121048e7557af235b584/test/inductor/test_perf.py#L810-L821 the fusion strategy might be suboptimal and it's good to have an override.

**Comment by lezcano** [comment]
That IMA is not super relevant really. I'd like to see how it affects the rest of torchbench tho.

**Review by peterbell10 — APPROVED** [review]
